File: ios-im-ui/Sources/FlareIMUI/EmojiSticker/EmojiStickerCatalog.swift
import Foundation
import ImageIO
import SwiftUI

#if canImport(UIKit)
import UIKit
public typealias FlarePlatformImage = UIImage
#elseif canImport(AppKit)
import AppKit
public typealias FlarePlatformImage = NSImage
#endif

/// Decoded animated-webp frames + per-frame durations (empty durations ⇒ static).
public struct FlareAnimatedFrames {
    public let frames: [FlarePlatformImage]
    public let durations: [Double]
    public var isAnimated: Bool { frames.count > 1 }
}

/// Decodes every frame of a (possibly animated) webp via ImageIO — no third-party
/// dependency. Returns nil if the file can't be read.
public func flareDecodeAnimatedWebp(url: URL?) -> FlareAnimatedFrames? {
    guard let url, let source = CGImageSourceCreateWithURL(url as CFURL, nil) else { return nil }
    let count = CGImageSourceGetCount(source)
    guard count > 0 else { return nil }
    var frames: [FlarePlatformImage] = []
    var durations: [Double] = []
    for i in 0..<count {
        guard let cg = CGImageSourceCreateImageAtIndex(source, i, nil) else { continue }
        #if canImport(UIKit)
        frames.append(UIImage(cgImage: cg))
        #else
        frames.append(NSImage(cgImage: cg, size: CGSize(width: cg.width, height: cg.height)))
        #endif
        var delay = 0.1
        if let props = CGImageSourceCopyPropertiesAtIndex(source, i, nil) as? [CFString: Any],
           let webp = props[kCGImagePropertyWebPDictionary] as? [CFString: Any] {
            if let d = webp[kCGImagePropertyWebPUnclampedDelayTime] as? Double, d > 0 {
                delay = d
            } else if let d = webp[kCGImagePropertyWebPDelayTime] as? Double, d > 0 {
                delay = d
            }
        }
        durations.append(delay)
    }
    return frames.isEmpty ? nil : FlareAnimatedFrames(frames: frames, durations: durations)
}

/// One sticker pack from the manifest.
public struct FlareStickerPack: Sendable, Identifiable {
    public let id: String       // protocol packageId (e.g. `gifs`, `classic`)
    public let dir: String      // e.g. `stickers/default`
    public let title: String
    public let stickerIds: [String]
}

/// Cross-platform emoji-pack + sticker catalog, backed by the flare-im-design
/// manifest bundled with this package (a committed mirror of the single source
/// `flare-im-design/assets/emoji-sticker`; see sync-resources.sh).
///
/// Resources load synchronously from `Bundle.module` on first access, so views
/// stay pure. Message views resolve assets by path convention; the picker + the
/// localized labels use the loaded manifest.
public final class FlareEmojiStickerCatalog: @unchecked Sendable {
    public static let shared = FlareEmojiStickerCatalog()

    public static let resourceRoot = "emoji-sticker"
    public static let stickerPackageGifs = "gifs"

    private let lock = NSLock()
    private var loaded = false
    public private(set) var emojiKeys: [String] = []
    private var emojiKeySet: Set<String> = []
    public private(set) var stickerPacks: [FlareStickerPack] = []
    private var locales: [String: [String: String]] = [:]
    private var imageCache: [String: FlarePlatformImage] = [:]

    private init() {}

    private func ensureLoaded() {
        lock.lock(); defer { lock.unlock() }
        if loaded { return }
        loaded = true

        guard let manifestURL = Self.resourceURL(name: "manifest", ext: "json", subdir: Self.resourceRoot),
              let data = try? Data(contentsOf: manifestURL),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return }

        if let emoji = json["emoji"] as? [String: Any], let keys = emoji["keys"] as? [String] {
            emojiKeys = keys
            emojiKeySet = Set(keys)
        }
        if let packs = json["stickerPacks"] as? [[String: Any]] {
            stickerPacks = packs.map { p in
                let items = (p["items"] as? [[String: Any]]) ?? []
                return FlareStickerPack(
                    id: p["id"] as? String ?? "",
                    dir: p["dir"] as? String ?? "",
                    title: p["title"] as? String ?? "",
                    stickerIds: items.compactMap { $0["id"] as? String }
                )
            }
        }
        if let localesURL = Self.resourceURL(name: "emoji-locales", ext: "json", subdir: Self.resourceRoot),
           let ldata = try? Data(contentsOf: localesURL),
           let ljson = try? JSONSerialization.jsonObject(with: ldata) as? [String: [String: String]] {
            locales = ljson
        }
    }

    public var isLoaded: Bool { ensureLoaded(); return loaded }

    public func loadedEmojiKeys() -> [String] { ensureLoaded(); return emojiKeys }
    public func loadedStickerPacks() -> [FlareStickerPack] { ensureLoaded(); return stickerPacks }

    public func hasEmojiKey(_ key: String) -> Bool {
        ensureLoaded()
        return emojiKeySet.contains(key.trimmingCharacters(in: .whitespaces))
    }

    /// On-disk sticker subdir for a protocol packageId (`gifs` → `default`).
    public static func stickerSubdir(forPackageId packageId: String?) -> String {
        let p = (packageId ?? "").trimmingCharacters(in: .whitespaces)
        return (p.isEmpty || p == stickerPackageGifs) ? "default" : p
    }

    private static func resourceURL(name: String, ext: String, subdir: String) -> URL? {
        Bundle.module.url(forResource: name, withExtension: ext, subdirectory: subdir)
    }

    /// Guards asset lookups against path traversal / injection from untrusted keys/ids.
    private static func isSafeComponent(_ value: String) -> Bool {
        !value.isEmpty && value.range(of: "^[A-Za-z0-9_-]+$", options: .regularExpression) != nil
    }

    public func emojiImageURL(_ key: String) -> URL? {
        let k = key.trimmingCharacters(in: .whitespaces)
        guard Self.isSafeComponent(k) else { return nil }
        return Self.resourceURL(name: k, ext: "webp", subdir: "\(Self.resourceRoot)/emoji")
    }

    public func stickerImageURL(stickerId: String, packageId: String?) -> URL? {
        let sid = stickerId.trimmingCharacters(in: .whitespaces)
        let subdir = Self.stickerSubdir(forPackageId: packageId)
        guard Self.isSafeComponent(sid), Self.isSafeComponent(subdir) else { return nil }
        return Self.resourceURL(name: sid, ext: "webp", subdir: "\(Self.resourceRoot)/stickers/\(subdir)")
    }

    /// Localized emoji-pack label; falls back to the raw key.
    public func emojiLabel(_ key: String, locale: String? = nil) -> String {
        ensureLoaded()
        let k = key.trimmingCharacters(in: .whitespaces)
        if k.isEmpty || locales.isEmpty { return k }
        let column = (locale ?? "en").lowercased().hasPrefix("zh") ? "zh-Hans" : "en"
        if let v = locales[column]?[k], !v.isEmpty { return v }
        if let v = locales["en"]?[k], !v.isEmpty { return v }
        return k
    }

    public func emojiBracketLabel(_ key: String, locale: String? = nil) -> String {
        "[\(emojiLabel(key, locale: locale))]"
    }

    /// Loads (and caches) a static platform image from a bundle URL.
    public func image(at url: URL?) -> FlarePlatformImage? {
        guard let url else { return nil }
        let key = url.path
        lock.lock()
        if let cached = imageCache[key] { lock.unlock(); return cached }
        lock.unlock()
        #if canImport(UIKit)
        let img = UIImage(contentsOfFile: url.path)
        #else
        let img = NSImage(contentsOfFile: url.path)
        #endif
        if let img {
            lock.lock(); imageCache[key] = img; lock.unlock()
        }
        return img
    }
}

extension Image {
    /// A SwiftUI `Image` from a platform image (cross-platform init).
    init(flarePlatformImage image: FlarePlatformImage) {
        #if canImport(UIKit)
        self.init(uiImage: image)
        #else
        self.init(nsImage: image)
        #endif
    }
}

File: android-im-ui/src/main/kotlin/com/flare/im/ui/EmojiStickerCatalog.kt
package com.flare.im.ui

import android.content.Context
import android.os.Build
import coil.ImageLoader
import coil.decode.GifDecoder
import coil.decode.ImageDecoderDecoder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject

/** One sticker pack from the manifest. */
data class FlareStickerPack(
    /** Protocol packageId (e.g. `gifs`, `classic`). */
    val id: String,
    /** On-disk dir relative to the resource root, e.g. `stickers/default`. */
    val dir: String,
    val title: String,
    val stickerIds: List<String>,
)

/**
 * Cross-platform emoji-pack + sticker catalog, backed by the flare-im-design
 * manifest bundled with this library (a symlink mirror of the single source
 * `flare-im-design/assets/emoji-sticker`, packaged under `assets/emoji-sticker`).
 *
 * Message views resolve assets by path convention (Coil renders the webp, a
 * missing file triggers the fallback). The picker + localized labels use
 * [ensureLoaded] to read the manifest + locales from `assets`.
 */
object FlareEmojiStickerCatalog {
    const val ASSET_ROOT = "emoji-sticker"

    /** Protocol packageId whose on-disk dir is `default/`. */
    const val STICKER_PACKAGE_GIFS = "gifs"

    @Volatile
    private var loaded = false
    val isLoaded: Boolean get() = loaded

    var emojiKeys: List<String> = emptyList()
        private set
    private var emojiKeySet: Set<String> = emptySet()

    var stickerPacks: List<FlareStickerPack> = emptyList()
        private set

    // locale column -> (key -> label)
    private var locales: Map<String, Map<String, String>> = emptyMap()

    /** Loads the manifest + locale labels once. Safe to call repeatedly. */
    suspend fun ensureLoaded(context: Context) = withContext(Dispatchers.IO) {
        if (loaded) return@withContext
        val assets = context.applicationContext.assets

        val manifest = JSONObject(assets.open("$ASSET_ROOT/manifest.json").use { it.readBytes().decodeToString() })
        val emoji = manifest.optJSONObject("emoji")
        val keysJson = emoji?.optJSONArray("keys")
        val keys = buildList {
            if (keysJson != null) for (i in 0 until keysJson.length()) add(keysJson.getString(i))
        }
        emojiKeys = keys
        emojiKeySet = keys.toSet()

        val packsJson = manifest.optJSONArray("stickerPacks")
        stickerPacks = buildList {
            if (packsJson != null) for (i in 0 until packsJson.length()) {
                val p = packsJson.getJSONObject(i)
                val itemsJson = p.optJSONArray("items")
                val ids = buildList {
                    if (itemsJson != null) for (j in 0 until itemsJson.length()) {
                        add(itemsJson.getJSONObject(j).getString("id"))
                    }
                }
                add(
                    FlareStickerPack(
                        id = p.optString("id"),
                        dir = p.optString("dir"),
                        title = p.optString("title"),
                        stickerIds = ids,
                    ),
                )
            }
        }

        locales = runCatching {
            val raw = JSONObject(assets.open("$ASSET_ROOT/emoji-locales.json").use { it.readBytes().decodeToString() })
            buildMap {
                for (col in raw.keys()) {
                    val map = raw.getJSONObject(col)
                    put(col, buildMap { for (k in map.keys()) put(k, map.getString(k)) })
                }
            }
        }.getOrDefault(emptyMap())

        loaded = true
    }

    fun hasEmojiKey(key: String): Boolean = emojiKeySet.contains(key.trim())

    /** On-disk sticker subdir for a protocol packageId (`gifs` → `default`). */
    fun stickerSubdirForPackageId(packageId: String?): String {
        val p = packageId?.trim().orEmpty()
        return if (p.isEmpty() || p == STICKER_PACKAGE_GIFS) "default" else p
    }

    /** Coil model for a bundled emoji asset. */
    fun emojiAssetUri(key: String): String =
        "file:///android_asset/$ASSET_ROOT/emoji/${key.trim()}.webp"

    fun stickerAssetUri(stickerId: String, packageId: String?): String =
        "file:///android_asset/$ASSET_ROOT/stickers/${stickerSubdirForPackageId(packageId)}/${stickerId.trim()}.webp"

    /** Localized emoji-pack label; falls back to the raw key. */
    fun emojiLabel(key: String, locale: String? = null): String {
        val k = key.trim()
        if (k.isEmpty() || locales.isEmpty()) return k
        val column = if ((locale ?: "en").lowercase().startsWith("zh")) "zh-Hans" else "en"
        locales[column]?.get(k)?.takeIf { it.isNotBlank() }?.let { return it.trim() }
        locales["en"]?.get(k)?.takeIf { it.isNotBlank() }?.let { return it.trim() }
        return k
    }

    fun emojiBracketLabel(key: String, locale: String? = null): String = "[${emojiLabel(key, locale)}]"
}

/** Coil [ImageLoader] that animates webp (emoji/stickers) on API 28+. */
fun flareEmojiStickerImageLoader(context: Context): ImageLoader =
    ImageLoader.Builder(context)
        .components {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                add(ImageDecoderDecoder.Factory())
            } else {
                add(GifDecoder.Factory())
            }
        }
        .build()
